File: src/test/java/quebec/virtualite/backend/services/rest/RestServerTest.java
package quebec.virtualite.backend.services.rest;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.slf4j.Logger;
import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;
import quebec.virtualite.backend.services.domain.DomainService;
import quebec.virtualite.backend.services.domain.entities.WheelEntity;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.springframework.http.HttpStatus.NOT_FOUND;
import static org.springframework.test.util.ReflectionTestUtils.setField;
import static quebec.virtualite.backend.TestConstants.ID;
import static quebec.virtualite.backend.TestConstants.NAME;
import static quebec.virtualite.backend.TestConstants.WHEEL;
import static quebec.virtualite.backend.TestConstants.WHEEL_DTO;
import static quebec.virtualite.backend.TestConstants.WHEEL_WITH_ID;
import static quebec.virtualite.utils.CollectionUtils.list;

@RunWith(MockitoJUnitRunner.class)
public class RestServerTest
{
    private static final String NEW_BRAND = "new brand";
    private static final String NEW_NAME = "new name";

    @InjectMocks
    private RestServer server;

    @Mock
    private DomainService mockedDomainService;

    @Mock
    private Logger mockedLogger;

    @Before
    public void before()
    {
        setField(server, "log", mockedLogger);
    }

    @Test
    public void addWheel()
    {
        // When
        server.addWheel(WHEEL_DTO);

        // Then
        verify(mockedDomainService).addWheel(WHEEL);
    }

    @Test
    public void deleteWheel()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.of(WHEEL_WITH_ID));

        // When
        server.deleteWheel(NAME);

        // Then
        verify(mockedDomainService).getWheel(NAME);
        verify(mockedDomainService).deleteWheel(WHEEL_WITH_ID);
    }

    @Test
    public void deleteWheel_whenNotFound()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.empty());

        // When
        Throwable exception = catchThrowable(() ->
            server.deleteWheel(NAME));

        // Then
        verify(mockedDomainService, never()).deleteWheel(any(WheelEntity.class));

        assertStatus(exception, NOT_FOUND);
    }

    @Test
    public void getWheelDetails()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.of(WHEEL_WITH_ID));

        // When
        WheelDTO response = server.getWheelDetails(NAME);

        // Then
        verify(mockedDomainService).getWheel(NAME);

        assertThat(response).isEqualTo(WHEEL_DTO);
    }

    @Test
    public void getWheelDetails_whenNotFound()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.empty());

        // When
        Throwable exception = catchThrowable(() ->
            server.getWheelDetails(NAME));

        // Then
        assertStatus(exception, NOT_FOUND);
    }

    @Test
    public void getWheelsDetails()
    {
        // Given
        given(mockedDomainService.getWheels())
            .willReturn(list(WHEEL_WITH_ID));

        // When
        List<WheelDTO> response = server.getWheelsDetails();

        // Then
        verify(mockedDomainService).getWheels();

        assertThat(response).isEqualTo(list(WHEEL_DTO));
    }

    @Test
    public void updateWheel()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.of(WHEEL_WITH_ID));

        // When
        server.updateWheel(NAME,
            new WheelDTO(NEW_BRAND, NEW_NAME));

        // Then
        verify(mockedDomainService).getWheel(NAME);
        verify(mockedDomainService).updateWheel(
            new WheelEntity(ID, NEW_BRAND, NEW_NAME));
    }

    @Test
    public void updateWheel_whenNotFound()
    {
        // Given
        given(mockedDomainService.getWheel(NAME))
            .willReturn(Optional.empty());

        // When
        Throwable exception = catchThrowable(() ->
            server.updateWheel(NAME, WHEEL_DTO));

        // Then
        assertStatus(exception, NOT_FOUND);
    }

    private static void assertStatus(Throwable exception, HttpStatus expectedStatus)
    {
        assertThat(exception)
            .isInstanceOf(ResponseStatusException.class)
            .hasFieldOrPropertyWithValue("status", expectedStatus);
    }
}

File: src/test/java/quebec/virtualite/backend/services/rest/RestServerTest.kt
package quebec.virtualite.backend.services.rest

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension
import org.slf4j.Logger
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.test.util.ReflectionTestUtils.setField
import quebec.virtualite.backend.TestConstants.BAD_CITY_DTO
import quebec.virtualite.backend.TestConstants.CITY
import quebec.virtualite.backend.TestConstants.CITY2
import quebec.virtualite.backend.TestConstants.CITY_DTO
import quebec.virtualite.backend.TestConstants.CITY_DTO2
import quebec.virtualite.backend.TestConstants.CITY_WITH_ID
import quebec.virtualite.backend.TestConstants.ID
import quebec.virtualite.backend.TestConstants.NAME
import quebec.virtualite.backend.TestConstants.NAME2
import quebec.virtualite.backend.TestConstants.PROVINCE2
import quebec.virtualite.backend.services.domain.DomainService
import quebec.virtualite.backend.services.domain.entities.CityEntity
import quebec.virtualite.backend.services.rest.impl.RestServer
import quebec.virtualite.backend.services.utils.TestUtils.assertInvalid
import quebec.virtualite.backend.services.utils.TestUtils.assertStatus
import quebec.virtualite.backend.services.utils.TestUtils.assertValid

@ExtendWith(MockitoExtension::class)
class RestServerTest
{
    @InjectMocks
    private lateinit var server: RestServer

    @Mock
    private lateinit var mockedDomainService: DomainService

    @Mock
    private lateinit var mockedLogger: Logger

    @BeforeEach
    fun before()
    {
        setField(server, "log", mockedLogger)
    }

    @Test
    fun addCity()
    {
        // When
        server.addCity(CITY_DTO)

        // Then
        verify(mockedDomainService).addCity(CITY)
    }

    @Test
    fun deleteCity()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(CITY)

        // When
        server.deleteCity(NAME)

        // Then
        verify(mockedDomainService).getCityDetails(NAME)
        verify(mockedDomainService).deleteCity(NAME)
    }

    @Test
    fun deleteCity_whenNotFound()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(null)

        // When
        val exception = catchThrowable {
            server.deleteCity(NAME)
        }

        // Then
        verify(mockedDomainService).getCityDetails(NAME)
        verify(mockedDomainService, never()).deleteCity(NAME)

        assertStatus(exception, NOT_FOUND)
    }

    @Test
    fun getCityDetails()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(CITY)

        // When
        val response = server.getCityDetails(NAME)

        // Then
        verify(mockedDomainService).getCityDetails(NAME)

        assertThat(response).isEqualTo(CITY_DTO)
    }

    @Test
    fun getCitiesDetails()
    {
        // Given
        given(mockedDomainService.getCitiesDetails())
            .willReturn(listOf(CITY, CITY2))

        // When
        val response = server.getCitiesDetails()

        // Then
        verify(mockedDomainService).getCitiesDetails()

        assertThat(response).isEqualTo(
            listOf(
                CITY_DTO,
                CITY_DTO2
            )
        )
    }

    @Test
    fun getCityDetails_whenNotFound()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(null)

        // When
        val exception = catchThrowable {
            server.getCityDetails(NAME)
        }

        // Then
        assertStatus(exception, NOT_FOUND)
    }

    @Test
    fun updateCity()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(CITY_WITH_ID)

        // When
        server.updateCity(NAME, CITY_DTO2)

        // Then
        verify(mockedDomainService).getCityDetails(NAME)
        verify(mockedDomainService).updateCity(CityEntity(ID, NAME2, PROVINCE2))
    }

    @Test
    fun updateCity_whenNotFound()
    {
        // Given
        given(mockedDomainService.getCityDetails(NAME))
            .willReturn(null)

        // When
        val exception = catchThrowable {
            server.updateCity(NAME, CITY_DTO)
        }

        // Then
        assertStatus(exception, NOT_FOUND)
    }

    @Test
    fun validate()
    {
        assertValid(server, "addCity", CITY_DTO)
        assertInvalid(server, "addCity", BAD_CITY_DTO)
        assertInvalid(server, "addCity", null)
        assertInvalid(server, "addCity", "")
        assertInvalid(server, "addCity", 10f)

        assertValid(server, "deleteCity", NAME)
        assertInvalid(server, "deleteCity", null)
        assertInvalid(server, "deleteCity", "")
        assertInvalid(server, "deleteCity", 10f)

        assertValid(server, "getCityDetails", NAME)
        assertInvalid(server, "getCityDetails", null)
        assertInvalid(server, "getCityDetails", "")
        assertInvalid(server, "getCityDetails", 10f)

        assertValid(server, "updateCity", NAME, CITY_DTO)
        assertInvalid(server, "updateCity", NAME, BAD_CITY_DTO)
        assertInvalid(server, "updateCity", "", BAD_CITY_DTO)
        assertInvalid(server, "updateCity", null, CITY_DTO)
        assertInvalid(server, "updateCity", "", CITY_DTO)
        assertInvalid(server, "updateCity", 10f, CITY_DTO)
    }
}
